lib/proxy/request_id_middleware.rb

```diff
@@ -0,0 +1,14 @@
+module Proxy
+  class RequestIdMiddleware
+    def initialize(app)
+      @app = app
+    end
+
+    def call(env)
+      Thread.current.thread_variable_set(:request_id, env['HTTP_X_REQUEST_ID']) if env.has_key?('HTTP_X_REQUEST_ID')
+      to_ret = @app.call(env)
+      Thread.current.thread_variable_set(:request_id, nil)
```

Comment:
Thanks for looking into this -- I didn't think to look into BodyProxy, but it explains why it's always one of the last thing to get logged. I also think using BodyProxy to clear the thread variable is probably the best approach.